### PR TITLE
Cleanup repository settings definitions - fs,url,s3

### PR DESCRIPTION
**Pull request by smakalias**
Cleaning up the repo settings definitions to use existing definitions whenever possible.

## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [ ] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 1]
```diff
@@ -126,9 +127,9 @@
 
     /**
      * When set to true metadata files are stored in compressed format. This setting doesn’t affect index
-     * files that are already compressed by default. Defaults to false.
+     * files that are already compressed by default. Defaults to true.
      */
-    static final Setting<Boolean> COMPRESS_SETTING = Setting.boolSetting("compress", false);
+    static final Setting<Boolean> COMPRESS_SETTING = Setting.boolSetting("compress", true); // TODO: in ES, it defaults to false!
```
Comment (seut) [inline es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: why do we want to change the defaults in a cleanup commit? please elaborate the TODO (what to do)

Comment (smakalias) [reply es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: please see #8636 - the default is also changed in upstream - will remove all todos

File: es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 2]
```diff
@@ -143,6 +144,19 @@
      */
     static final Setting<String> BASE_PATH_SETTING = Setting.simpleString("base_path");
 
+    public static List<Setting> optionalSettingsToValidate() {
```
Comment (kovrus) [inline es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: why those are called optionalSettings?

File: sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java [hunk 3]
```diff
@@ -138,4 +98,28 @@ protected void configure() {
...
+     * with it's key suffix acting as the new key
+     *
+     * eg. From Setting: s3.client.default.endpoint
+     * a new Setting will be created with same characteristics
+     * and new key name: endpoint
+     */
+    private static List<Setting> renameSettingsUsingSuffixAsKey(List<Setting> settingList) {
+        return settingList
+            .stream()
+            .map(s -> s.copyAndRename(k -> getSuffixOrInput((String) k)))
```
Comment (kovrus) [inline sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java]: string cast should not be needed here

File: sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java [hunk 4]
```diff
@@ -50,29 +49,19 @@
...
-            .put("path", Setting.simpleString("path", Setting.Property.NodeScope))
-            .put("load_defaults", Setting.boolSetting("load_defaults", true, Setting.Property.NodeScope))
-            .put("concurrent_streams", Setting.intSetting("concurrent_streams", 5, Setting.Property.NodeScope))
-            .put("compress", Setting.boolSetting("compress", true, Setting.Property.NodeScope))
-            // We cannot use a ByteSize setting as it doesn't support NULL and it must be NULL as default to indicate to
-            // not override the default behaviour.
-            .put("chunk_size", Setting.simpleString("chunk_size"))
-            .build()) {
+        Map.of(),
+        // tmp comment for review : left out 'concurrent_streams' as I didn't see any use of it
```
Comment (kovrus) [inline sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java]: maybe we should still keep it as it is unrelated change and fix or remove it in other pr?

File: sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java [hunk 5]
```diff
@@ -50,29 +49,19 @@
...
+        convertSettingListToMap(HdfsRepository.settingsToValidate())) {
```
Comment (kovrus) [inline sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java]: are these optional? if so we probably have to name the method accordingly to be consistent with the TypeSettings initialized above.

File: es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java (hunk 2 above)
Comment (smakalias) [reply es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: These are considered optional for the parameters provided at the `create repository` statement. 
See https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/analyze/repositories/TypeSettings.java#L44

I'm open to other suggestions.

Comment (kovrus) [reply es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: i was asking it coz the settings list has ACCESS_KEY_SETTING and SECRET_KEY_SETTING, but i see

File: sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java (hunk 4 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java]: good point. it's now handled here: #8671

Comment (seut) [reply sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java]: yes this should be done in a dedicated commit/PR

File: es/es-server/src/main/java/org/elasticsearch/common/settings/Setting.java [hunk 6]
```diff
@@ -653,6 +654,18 @@ public AffixSetting(AffixKey key, Setting<T> delegate, Function<String, Setting<
...
 
+        /**
+         * Returns the delegate setting using the key suffix as the new setting key
+         * This is meant to serve as a convenience method to extract a setting
+         * of the same type as it's delegate.
+         * Use {@link #getConcreteSetting} or {@link #getConcreteSettingForNamespace(String)}
+         * to retrieve valid concrete settings
+         *
+         */
+        public Setting<T> asSettingUsingSuffixAsKey() {
```
Comment (seut) [inline es/es-server/src/main/java/org/elasticsearch/common/settings/Setting.java]: unused?

File: sql/src/test/java/io/crate/analyze/RepositoryParamValidatorTest.java [hunk 7]
```diff
@@ -91,6 +92,14 @@ public void testHdfsSecurityPrincipal() throws Exception {
         assertThat(settings.get("uri"), is("hdfs://ha-name:8020"));
     }
 
+    @Test
+    public void testHdfsUriDefaultValueIsNull() throws Exception {
+        GenericProperties genericProperties = new GenericProperties();
+        genericProperties.add(new GenericProperty("path", new StringLiteral("/user/name/data")));
```
Comment (seut) [inline sql/src/test/java/io/crate/analyze/RepositoryParamValidatorTest.java]: `GenericProperties.EMPTY` can be used instead (simplify test case)

File: sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java [hunk 8]
```diff
@@ -138,4 +97,28 @@ protected void configure() {
         typeSettingsBinder.addBinding(HDFS).toInstance(HDFS_SETTINGS);
         typeSettingsBinder.addBinding(S3).toInstance(S3_SETTINGS);
     }
+
+    private static Map<String, Setting> convertSettingListToMap(List<Setting> settingList) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java]:
```suggestion
    private static Map<String, Setting> groupSettingsByKey(List<Setting> settings) {
```

Otherwise you'll have to read the implementation to know what it does exactly - making this abstraction more complex than useful

File: sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java [hunk 9]
```diff
@@ -138,4 +97,28 @@ protected void configure() {
...
+    /**
+     * Copy and rename {@link Setting}s.
+     * For each Setting, a new Setting is created
+     * with it's key suffix acting as the new key
+     *
+     * eg. From Setting: s3.client.default.endpoint
+     * a new Setting will be created with same characteristics
+     * and new key name: endpoint
+     */
+    private static List<Setting> renameSettingsUsingSuffixAsKey(List<Setting> settingList) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java]:
```suggestion
    private static List<Setting> renameSettingsUsingSuffixAsKey(List<Setting> settings) {
```

File: es/es-server/src/main/java/org/elasticsearch/repositories/fs/FsRepository.java [hunk 10]
```diff
@@ -62,6 +63,14 @@
     public static final Setting<ByteSizeValue> REPOSITORIES_CHUNK_SIZE_SETTING = Setting.byteSizeSetting("repositories.fs.chunk_size",
         new ByteSizeValue(Long.MAX_VALUE), new ByteSizeValue(5), new ByteSizeValue(Long.MAX_VALUE), Property.NodeScope);
 
+    public static List<Setting> mandatorySettingsToValidate() {
```
Comment (mfussenegger) [inline es/es-server/src/main/java/org/elasticsearch/repositories/fs/FsRepository.java]:
```suggestion
    public static List<Setting> mandatorySettings() {
```

File: es/es-server/src/main/java/org/elasticsearch/repositories/fs/FsRepository.java [hunk 11]
```diff
@@ -62,6 +63,14 @@
...
+        return List.of(LOCATION_SETTING);
+    }
+
+    public static List<Setting> optionalSettingsToValidate() {
```
Comment (mfussenegger) [inline es/es-server/src/main/java/org/elasticsearch/repositories/fs/FsRepository.java]:
```suggestion
    public static List<Setting> optionalSettings() {
```

The `toValidate` part implies a very specific use-case, that is also kind of redundant.

File: es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java (hunk 2 above)
Comment (mfussenegger) [inline es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]:
```suggestion
    public static List<Setting> optionalSettings() {
```

File: es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3ClientSettings.java [hunk 12]
```diff
@@ -100,6 +101,13 @@
     static final Setting<Boolean> USE_THROTTLE_RETRIES_SETTING = Setting.boolSetting(DEFAULT_PREFIX + "use_throttle_retries",
                                                                                      ClientConfiguration.DEFAULT_THROTTLE_RETRIES,
                                                                                      Property.NodeScope);
+    public static List<Setting> optionalSettingsToValidate() {
```
Comment (mfussenegger) [inline es/es-repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3ClientSettings.java]:
```suggestion
    public static List<Setting> optionalSettings() {
```

File: es/es-repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java [hunk 13]
```diff
@@ -47,32 +48,56 @@
...
+
+    private static final Setting<String> URI_SETTING = Setting.simpleString("uri", Setting.Property.NodeScope);
+
+    private static final Setting<String> PATH_SETTING = Setting.simpleString("path", Setting.Property.NodeScope);
+
+    private static final Setting<String> SECURITY_PRINCIPAL_SETTING = Setting.simpleString("security.principal", Setting.Property.NodeScope);
+
+    private static final Setting<Boolean> LOAD_DEFAULTS_SETTING = Setting.boolSetting("load_defaults",true, Setting.Property.NodeScope);
+
+    public static List<Setting> optionalSettingsToValidate() {
```
Comment (mfussenegger) [inline es/es-repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java]:
```suggestion
    public static List<Setting> optionalSettings() {
```

File: es/es-repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java [hunk 14]
```diff
@@ -47,32 +48,56 @@
...
+    // buffer size passed to HDFS read/write methods
+    // TODO: why 100KB?
+    private static final ByteSizeValue DEFAULT_BUFFER_SIZE = new ByteSizeValue(100, ByteSizeUnit.KB);
+
+    private static final Setting<ByteSizeValue> BUFFER_SIZE_SETTING =
+        Setting.byteSizeSetting("buffer_size", DEFAULT_BUFFER_SIZE);
+
+    // We cannot use a ByteSize setting as it doesn't support NULL
+    // and it must be NULL as default to indicate to not override the default behaviour.
+    private static final Setting<String> CHUNK_SIZE_SETTING = Setting.simpleString("chunk_size",  Setting.Property.NodeScope);
```
[the remark this answers is not in the document]
Comment (smakalias) [reply es/es-repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java]: removed all hdfs-related changes for now - thus this is not relevant anymore.

File: es/es-repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java [hunk 15]
```diff
@@ -47,32 +48,56 @@
...
+
+    private static final Setting<String> URI_SETTING = Setting.simpleString("uri", Setting.Property.NodeScope);
```
[the remark this answers is not in the document]
Comment (smakalias) [reply es/es-repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java]: removed all hdfs-related changes for now - thus this is not relevant anymore.

File: es/es-repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java [hunk 16]
```diff
@@ -47,32 +48,56 @@
...
+
+    private static final Setting<String> PATH_SETTING = Setting.simpleString("path", Setting.Property.NodeScope);
+
+    private static final Setting<String> SECURITY_PRINCIPAL_SETTING = Setting.simpleString("security.principal", Setting.Property.NodeScope);
+
+    private static final Setting<Boolean> LOAD_DEFAULTS_SETTING = Setting.boolSetting("load_defaults",true, Setting.Property.NodeScope);
```
[the remark this answers is not in the document]
Comment (smakalias) [reply es/es-repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java]: removed all hdfs-related changes for now - thus this is not relevant anymore.

File: es/es-repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java [hunk 17]
```diff
@@ -47,32 +48,56 @@
...
-    // TODO: why 100KB?
-    private static final ByteSizeValue DEFAULT_BUFFER_SIZE = new ByteSizeValue(100, ByteSizeUnit.KB);
 
     public HdfsRepository(RepositoryMetaData metadata, Environment environment,
                           NamedXContentRegistry namedXContentRegistry) {
         super(metadata, environment.settings(), namedXContentRegistry);
 
         this.environment = environment;
-        this.chunkSize = metadata.settings().getAsBytesSize("chunk_size", null);
+        this.chunkSize = metadata.settings().getAsBytesSize(CHUNK_SIZE_SETTING.getKey(), null);
```
[the remark this answers is not in the document]
Comment (smakalias) [reply es/es-repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsRepository.java]: removed all hdfs-related changes for now - thus this is not relevant anymore.

File: sql/src/test/java/io/crate/analyze/RepositoryParamValidatorTest.java (hunk 7 above)
Comment (smakalias) [reply sql/src/test/java/io/crate/analyze/RepositoryParamValidatorTest.java]: removed all hdfs-related changes for now - thus this is not relevant anymore.

File: sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java (hunk 3 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java]: this was a surprise for me too - I didn't expect that a cast is in need as the `apply` method is templated.  
However, it expects an `Object` for some reason. Without the cast I get an error: "incompatible types: java.lang.Object cannot be converted to java.lang.String"

What am I missing here?

Comment (kovrus) [reply sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java]: type seemed to be inferred from Setting<T> :/

Comment (smakalias) [reply sql/src/main/java/io/crate/analyze/repositories/RepositorySettingsModule.java]: > type seemed to be inferred from Setting :/

yeah, I guess I could change all methods returning `List<Setting>` to return `List<Setting<T>>`
Not sure if it's worth it to be honest
